penpot_rock/rockcraft.yaml

```diff
@@ -0,0 +1,230 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+name: penpot
+summary: Web-based open-source design tool.
+description: Penpot is the web-based open-source design tool that bridges the gap between designers and developers.
+version: &penpot-version 2.1.1
+license: MPL-2.0
+
+base: ubuntu@24.04
+build-base: ubuntu@24.04
+platforms:
+  amd64:
+
+run-user: _daemon_
+
+parts:
+  frontend:
+    plugin: nil
+    source-type: git
+    source: https://github.com/penpot/penpot.git
+    source-tag: *penpot-version
+    stage-packages:
+      - nginx-full
+      - gettext-base
+    build-packages:
+      - openjdk-21-jdk
+      - rlwrap
+      - npm
+      - rsync
+    build-environment:
+      - VERSION: *penpot-version
+    override-build: |
+      craftctl default
+      
+      # install clojure
+      curl -L https://github.com/clojure/brew-install/releases/download/1.11.3.1463/linux-install.sh -o install-clojure
+      echo "0c41063a2fefb53a31bc1bc236899955f759c5103dc0495489cdd74bf8f114bb  install-clojure" | shasum -c
+      chmod +x install-clojure
+      ./install-clojure
+      
+      cd frontend
+
+      # install yarn
+      npm install --global corepack
+      corepack enable
+      corepack prepare yarn@4.2.2
+      
+      # build assets
+      export NODE_ENV=production
+      yarn install
+      clojure -M:dev:shadow-cljs release main --config-merge "{:release-version \"$(git rev-parse --short HEAD)-$(date +%s)\"}"
+      yarn run compile
+      sed -i -re "s/\%version\%/${VERSION}/g" ./resources/public/index.html
+      sed -i -re "s/\%buildDate\%/$(date -R);/g" ./resources/public/index.html
+      echo ${VERSION} > ./resources/public/version.txt
+      
+      mkdir -p $CRAFT_PART_INSTALL/var/www/
+      mkdir -p $CRAFT_PART_INSTALL/etc/nginx/
+      
+      cd ..
+      cp -r ./frontend/resources/public $CRAFT_PART_INSTALL/var/www/app
+      cp ./docker/images/files/nginx-mime.types $CRAFT_PART_INSTALL/etc/nginx/mime.types
+      cp ./docker/images/files/config.js $CRAFT_PART_INSTALL/var/www/app/js/config.js
+      cp ./docker/images/files/nginx.conf $CRAFT_PART_INSTALL/etc/nginx/nginx.conf.template
+      mkdir -p $CRAFT_PART_INSTALL/opt/penpot/frontend
+      cp ./docker/images/files/nginx-entrypoint.sh $CRAFT_PART_INSTALL/opt/penpot/frontend/nginx-entrypoint.sh
+      chmod +x $CRAFT_PART_INSTALL/opt/penpot/frontend/nginx-entrypoint.sh
+      
+      # nginx runs as non-root user
+      sed -i "s/listen 80 default_server;/listen 8080 default_server;/g" $CRAFT_PART_INSTALL/etc/nginx/nginx.conf.template
+      sed -i "s/pid \\/run\\/nginx.pid;/pid \\/opt\\/penpot\\/frontend\\/nginx.pid;/g" $CRAFT_PART_INSTALL/etc/nginx/nginx.conf.template
+      sed -i "s/user www-data;/user _daemon_;/g" $CRAFT_PART_INSTALL/etc/nginx/nginx.conf.template
+
+    override-stage: |
+      chown -R 584792:584792 $CRAFT_PART_INSTALL/etc/nginx/
```

Comment:
Yes, that's a good idea. I didn't know about this option. Updated, thanks!

Reply:
I tried using the `permissions` field, but it seems to have an issue with following symbolic links when changing permissions. Since the files are copied and changed one by one, sometimes the target is not copied, and at that time, change permission will fail. I think I might need to report this issue to the rockcraft team. But for now, I've just reverted to the old way of doing it.

Here are some references:
https://github.com/canonical/craft-parts/blob/a47e795a598eed7b98e1c4e19061677e1379adbc/craft_parts/executor/migration.py#L123-L127
https://github.com/canonical/craft-parts/blob/a47e795a598eed7b98e1c4e19061677e1379adbc/craft_parts/permissions.py#L80-L90